src/Lean/PrettyPrinter/Delaborator/Basic.lean

```diff
@@ -170,12 +188,40 @@ def withBindingBodyUnusedName {α} (d : Syntax → DelabM α) : DelabM α := do
 @[inline] def liftMetaM {α} (x : MetaM α) : DelabM α :=
   liftM x
 
+def addTermInfo (pos : Pos) (stx : Syntax) (e : Expr) (isBinder : Bool := false) : DelabM Unit := do
+  let info ← mkTermInfo stx e isBinder
+  modify fun s => { s with infos := s.infos.insert pos info }
+where
+  mkTermInfo stx e isBinder := do Info.ofTermInfo {
+    elaborator := `Delab,
+    stx := stx,
+    lctx := (← getLCtx),
+    expectedType? := none,
+    expr := e,
+    isBinder := isBinder
+ }
+
+def addFieldInfo (pos : Pos) (projName fieldName : Name) (stx : Syntax) (val : Expr) : DelabM Unit := do
+  let info ← mkFieldInfo projName fieldName stx val
+  modify fun s => { s with infos := s.infos.insert pos info }
+where
+  mkFieldInfo projName fieldName stx val := do Info.ofFieldInfo {
+    projName := projName,
+    fieldName := fieldName,
+    lctx := (← getLCtx),
+    val := val,
+    stx := stx
+  }
+
 partial def delabFor : Name → Delab
   | Name.anonymous => failure
-  | k              => do
-    (delabAttribute.getValues (← getEnv) k).firstM id >>= annotateCurPos <|>
-      -- have `app.Option.some` fall back to `app` etc.
-      delabFor k.getRoot
+  | k              =>
+    (do let stx ← (delabAttribute.getValues (← getEnv) k).firstM id
+        let stx ← annotateCurPos stx
+        addTermInfo (← getPos) stx (← getExpr)
```

Comment:
What is the overhead of using `addTermInfo` here? Is the slowdown significant? Just curious.

Reply:
I didn't test in a profiler, but in terms of wall time on both a long trace (~1s to print) and a long expression (~1.3s), I observed no difference with and without the `addTermInfo/addFieldInfo` calls.